dvc/repo/ls.py

```diff
@@ -0,0 +1,90 @@
+import os
+
+from dvc.exceptions import DvcException, OutputNotFoundError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    from dvc.external_repo import external_repo
+    from dvc.repo import Repo
+
+    with external_repo(url, rev) as repo:
+        result = set()
+        if isinstance(repo, Repo):
+            result.update(_ls_outs_repo(repo, target, recursive))
+        if not outputs_only:
+            try:
+                result.update(_ls_files_repo(repo, target, recursive))
+            except DvcException as exc:
+                if not len(result):
+                    raise exc
+        result = sorted(list(result))
+    return result
+
+
+def _ls_files_repo(repo, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+
+    def prettify(prefix, filename):
+        if not prefix == os.curdir:
+            filename = os.path.join(prefix, filename)
+        return filename
+
+    files = []
+    target_path = (
+        os.path.join(repo.root_dir, target) if target else repo.root_dir
+    )
+    tree = CleanTree(WorkingTree(target_path))
+
+    try:
+        if not os.path.exists(target_path):
+            raise FileNotFoundError("{} doesn't exist".format(target_path))
+
+        if target:
+            files.append(prettify(os.curdir, target))
+            if not recursive:
+                return files
+
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            prefix = os.path.relpath(dirpath, repo.root_dir)
+            files.extend(map(lambda f: prettify(prefix, f), dirnames))
+            files.extend(map(lambda f: prettify(prefix, f), filenames))
+            if not recursive:
+                break
+    except FileNotFoundError:
+        raise DvcException("target {} doesn't exist".format(target))
+    except NotADirectoryError:
```

Comment:
This one only applies to `tree.walk`, right? And `FileNotFoundError` re-raise is not needed as described in https://github.com/iterative/dvc/pull/3246/files#r376130721 , so you at least you should narrow the scope of this try &except NotADirectoryError to `tree.walk`, so this is less confusing.